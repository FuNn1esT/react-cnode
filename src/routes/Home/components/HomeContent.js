import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { List, Avatar, Card, Divider, Tag } from 'antd';
import { timeUtils } from 'utils';
import styles from './HomeContent.scss';

@connect(
  state => ({
    tab: state.basic.tab,
    loading: state.home.loading,
    topicsData: state.home.topicsData,
  }),
)
class HomeContent extends React.Component {
  static propTypes = {
    loading: PropTypes.bool.isRequired,
    topicsData: PropTypes.array,
  };

  renderTag = (key, isTop, isGood) => {
    const { tab } = this.props;
    const Tabs = {
      top: <Tag color="red">置顶</Tag>,
      good: <Tag color="green">精华</Tag>,
      share: <Tag color="blue">分享</Tag>,
      ask: <Tag color="orange">问答</Tag>,
      job: <Tag color="geekblue">招聘</Tag>,
    };
    if (isTop) {
      return Tabs.top;
    } else if (isGood && !Object.is(tab, 'good')) {
      return Tabs.good;
    } else if (Object.is(tab, 'all')) {
      return Tabs[key];
    } else {
      return null;
    }
  };

  render() {
    const { loading, topicsData } = this.props;
    return (
      <List
        className={styles.container}
        size="large"
        grid={{ gutter: 16, column: 1 }}
        dataSource={topicsData}
        loading={loading}
        renderItem={item => (
          <List.Item
            key={item.id}
          >
            <Card>
              <Card.Meta
                title={
                  <div className={styles.item}>
                    <div>
                      <Avatar
                        size="small"
                        shape="square"
                        src={item.author.avatar_url}
                      />
                      {item.author.loginname}
                    </div>
                    <div>
                      {this.renderTag(item.tab, item.top, item.good)}
                      <Link
                        to={{ pathname: `/topic/${item.id}` }}
                      >
                        {item.title}
                      </Link>
                    </div>
                  </div>
                }
                description={
                  <div>
                    <div className={styles.content}>
                      {item.content.replace(/<[^>]*>/g, '')}
                    </div>
                    <div>
                      {`阅读数:${item.visit_count}`}
                      <Divider type="vertical" />
                      {`${item.reply_count}条评论`}
                      <Divider type="vertical" />
                      {timeUtils.fromNow(item.last_reply_at)}
                    </div>
                  </div>
                }
              />
            </Card>
          </List.Item>
        )}
      />
    );
  }
}

export default HomeContent;
