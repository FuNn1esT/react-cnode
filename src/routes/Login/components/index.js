export { default as LoginToken } from './LoginToken';
export { default as LoginSubmit } from './LoginSubmit';
