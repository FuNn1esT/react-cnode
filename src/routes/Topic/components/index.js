export { default as TopicContent } from './TopicContent';
export { default as TopicReply } from './TopicReply';
export { default as TopicEditor } from './TopicEditor';
