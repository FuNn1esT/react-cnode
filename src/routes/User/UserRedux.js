import { UserService, TopicCollectService } from 'services';
import { createAction, handleActions } from 'redux-actions';

const LOAD_USER = 'routes/User/LOAD_USER';
const LOAD_USER_SUCCESS = 'routes/User/LOAD_USER_SUCCESS';
const LOAD_USER_ERROR = 'routes/User/LOAD_USER_ERROR';

const initialState = {
  loading: false,
  userData: {},
  error: false,
};

const loadUser = createAction(LOAD_USER);
const loadUserSuccess = createAction(
  LOAD_USER_SUCCESS,
  response => response,
);
const loadUserError = createAction(LOAD_USER_ERROR);

export const getUserData = name => async (dispatch) => {
  dispatch(loadUser());
  try {
    const { data: { data } } =
      await UserService.getUser(name);
    const { data: { data: collectData } } =
      await TopicCollectService.getUserCollect({ name });
    data.collect = collectData;
    dispatch(loadUserSuccess(data));
  } catch (e) {
    dispatch(loadUserError());
  }
};

const reducer = handleActions({
  [LOAD_USER]: state => ({
    ...state,
    loading: true,
    userData: {},
    error: false,
  }),

  [LOAD_USER_SUCCESS]: (state, action) => ({
    ...state,
    loading: false,
    error: false,
    userData: action.payload,
  }),

  [LOAD_USER_ERROR]: state => ({
    ...state,
    loading: false,
    error: true,
  }),
}, initialState);

export default reducer;
