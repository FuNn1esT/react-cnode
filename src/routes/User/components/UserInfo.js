import React from 'react';
import { Card, Avatar } from 'antd';
import { timeUtils } from 'utils';
import styles from './UserInfo.scss';

const UserInfo = ({ loading, user }) => (
  <Card className={styles.container} loading={loading}>
    <Card.Meta
      avatar={<Avatar src={user.avatar_url} />}
      title={user.loginname}
      description={
        <div>
          <div>{`Github: ${user.githubUsername}`}</div>
          <div>{`注册时间: ${timeUtils.fromNow(user.create_at)}`}</div>
        </div>
      }
    />
  </Card>
);

export default UserInfo;
