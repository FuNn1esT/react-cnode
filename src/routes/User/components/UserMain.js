import React from 'react';
import PropTypes from 'prop-types';
import { Card, List, Avatar } from 'antd';

const tabList = [
  {
    key: 'topics',
    tab: '最近创建的话题',
  }, {
    key: 'replies',
    tab: '最近参与的话题',
  },
];

export const UserMainContent = ({ data }) => {
  if (data.length === 0) {
    return <div>无话题</div>;
  } else {
    return (
      <List
        dataSource={data}
        renderItem={item => (
          <List.Item>
            <List.Item.Meta
              avatar={<Avatar src={item.author.avatar_url} />}
              title={<a href={`/topic/${item.id}`}>{item.title}</a>}
            />
          </List.Item>
        )}
      />
    );
  }
};

class UserMain extends React.Component {
  static propTypes = {
    loading: PropTypes.bool.isRequired,
    dataList: PropTypes.shape({
      topics: PropTypes.array.isRequired,
      replies: PropTypes.array.isRequired,
    }).isRequired,
  };

  state = {
    key: 'topics',
  };

  onTabChange = (key) => {
    this.setState({ key });
  };

  render() {
    const { loading, dataList } = this.props;
    const { key } = this.state;

    return (
      <Card
        loading={loading}
        tabList={tabList}
        onTabChange={this.onTabChange}
      >
        <UserMainContent data={dataList[key]} />
      </Card>
    );
  }
}

export default UserMain;
