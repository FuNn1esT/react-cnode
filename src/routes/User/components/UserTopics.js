import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { timeUtils } from 'utils';
import { Card, List, Avatar, Tag } from 'antd';
import styles from './UserTopics.scss';

const tabList = [
  {
    key: 'topics',
    tab: '最近创建的话题',
  },
  {
    key: 'replies',
    tab: '最近参与的话题',
  },
  {
    key: 'collect',
    tab: '收藏的话题',
  },
];

export const UserMainContent = ({ data }) => (
  <List
    locale={{ emptyText: '无话题' }}
    dataSource={data}
    renderItem={item => (
      <List.Item>
        <List.Item.Meta
          avatar={<Avatar src={item.author.avatar_url} />}
          title={<Link to={`/topic/${item.id}`}>{item.title}</Link>}
          description={
            <div>
              来自：<Tag color="purple">{item.author.loginname}</Tag>
            </div>
          }
        />
        {timeUtils.fromNow(item.last_reply_at)}
      </List.Item>
    )}
  />
);

class UserTopics extends React.Component {
  static propTypes = {
    dataList: PropTypes.object.isRequired,
  };

  state = {
    key: 'topics',
  };

  onTabChange = key => {
    this.setState({ key });
  };

  render() {
    const { dataList } = this.props;
    const { key } = this.state;

    return (
      <Card
        className={styles.container}
        tabList={tabList}
        defaultActiveTabKey="topics"
        onTabChange={this.onTabChange}
      >
        <UserMainContent data={dataList[key]} />
      </Card>
    );
  }
}

UserMainContent.propTypes = {
  data: PropTypes.array,
};

export default UserTopics;
