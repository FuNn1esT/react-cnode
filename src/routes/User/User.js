import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import classNames from 'classnames';
import Exception from 'components/Exception';
import { getUserData } from './UserRedux';
import { UserInfo, UserTopics } from './components';
import styles from './User.scss';

const cls = classNames(styles.mdContainer);

@connect(
  state => ({
    loading: state.getIn(['user', 'loading']),
    userData: state.getIn(['user', 'userData']),
    error: state.getIn(['user', 'error']),
  }),
  {
    getUserData,
  },
)
class User extends React.Component {
  componentDidMount() {
    const {
      match: {
        params: { name },
      },
    } = this.props;
    this.props.getUserData(name);
  }

  render() {
    const { loading, userData, error } = this.props;
    if (error) {
      return <Exception />;
    }
    return (
      <div className={cls}>
        <UserInfo loading={loading} user={userData} />
        {!loading && (
          <UserTopics
            dataList={{
              topics: userData.recent_topics || [],
              replies: userData.recent_replies || [],
              collect: userData.collect || [],
            }}
          />
        )}
      </div>
    );
  }
}

User.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      name: PropTypes.string,
    }),
  }).isRequired,

  loading: PropTypes.bool,
  userData: PropTypes.object,
  error: PropTypes.bool,
  getUserData: PropTypes.func,
};

export default User;
