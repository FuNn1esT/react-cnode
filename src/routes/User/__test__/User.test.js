import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { Map } from 'immutable';
import thunk from 'redux-thunk';
import toJson from 'enzyme-to-json';
import MockDate from 'mockdate';
import configureStore from 'redux-mock-store';
import User from '../User';
import { UserTopics } from '../components';

MockDate.set(new Date('2018-05-21T16:00:00.000Z'));

const middleware = [thunk];

const userProps = {
  match: {
    params: {
      name: 'admin',
    },
  },
};

const userInitialState = Map({
  user: Map({
    loading: false,
    userData: {
      avatar_url: '//',
      collect: [{}, {}],
      create_at: '2018-05-20T16:00:00.000Z',
      githubUsername: 'admin',
      loginname: 'admin',
      score: 520,
      recent_topics: [
        {
          author: {
            avatar_url: '//',
            loginname: 'admin',
          },
          id: '007',
          title: 'title',
          last_reply_at: '2018-05-20T16:00:00.000Z',
        },
      ],
      recent_replies: [],
    },
    error: false,
  }),
  getUserData: jest.fn(),
});

const dataListMock = {
  collect: [],
  replies: [],
  topics: [],
};

describe('<User />', () => {
  it('should render correctly', () => {
    const store = configureStore(middleware)(userInitialState);
    const tree = mount(
      <MemoryRouter>
        <User store={store} {...userProps} />
      </MemoryRouter>,
    );
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });

  it('should render <Exception />', () => {
    const initialState = userInitialState.setIn(['user', 'error'], true);
    const store = configureStore(middleware)(initialState);
    const tree = mount(
      <MemoryRouter>
        <User store={store} {...userProps} />
      </MemoryRouter>,
    );
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });
});

describe('<UserTopics />', () => {
  it('should change state when call onTableChange()', () => {
    const wrapper = shallow(<UserTopics dataList={dataListMock} />);
    wrapper.instance().onTabChange('topics');
    expect(wrapper.state().key).toEqual('topics');
  });
});
