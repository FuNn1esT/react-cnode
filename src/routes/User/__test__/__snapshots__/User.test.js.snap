// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<User /> should render <Exception /> 1`] = `
<div>
  <span>
    网络出小差了！请稍后再试
  </span>
</div>
`;

exports[`<User /> should render correctly 1`] = `
<div
  className=""
>
  <div
    className="ant-card ant-card-bordered"
  >
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <div
        className="ant-card-meta"
      >
        <div
          className="ant-card-meta-avatar"
        >
          <span
            className="ant-avatar ant-avatar-circle ant-avatar-image"
          >
            <img
              onError={[Function]}
              src="//"
            />
          </span>
        </div>
        <div
          className="ant-card-meta-detail"
        >
          <div
            className="ant-card-meta-title"
          >
            <div
              className="ant-tag ant-tag-purple"
              data-show={true}
              style={
                Object {
                  "backgroundColor": null,
                }
              }
            >
              admin
            </div>
          </div>
          <div
            className="ant-card-meta-description"
          >
            <div>
              Github：
              <a
                href="https://github.com/admin"
                rel="noopener noreferrer"
                target="_blank"
              >
                admin
              </a>
            </div>
            <div>
              积分：
              520
            </div>
            <div>
              收藏话题：
              2
               个
            </div>
            <div>
              注册时间：
              a day ago
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-card ant-card-bordered ant-card-contain-tabs"
  >
    <div
      className="ant-card-head"
    >
      <div
        className="ant-card-head-wrapper"
      />
      <div
        className="ant-tabs ant-tabs-top ant-card-head-tabs ant-tabs-large ant-tabs-line"
        style={Object {}}
      >
        <div
          className="ant-tabs-bar"
          onKeyDown={[Function]}
          role="tablist"
          tabIndex="0"
        >
          <div
            className="ant-tabs-nav-container"
          >
            <span
              className="ant-tabs-tab-prev ant-tabs-tab-btn-disabled"
              onClick={null}
              onTransitionEnd={[Function]}
              unselectable="unselectable"
            >
              <span
                className="ant-tabs-tab-prev-icon"
              />
            </span>
            <span
              className="ant-tabs-tab-next ant-tabs-tab-btn-disabled"
              onClick={null}
              unselectable="unselectable"
            >
              <span
                className="ant-tabs-tab-next-icon"
              />
            </span>
            <div
              className="ant-tabs-nav-wrap"
            >
              <div
                className="ant-tabs-nav-scroll"
              >
                <div
                  className="ant-tabs-nav ant-tabs-nav-animated"
                >
                  <div
                    className="ant-tabs-ink-bar ant-tabs-ink-bar-animated"
                  />
                  <div
                    aria-disabled="false"
                    aria-selected="true"
                    className="ant-tabs-tab-active ant-tabs-tab"
                    onClick={[Function]}
                    role="tab"
                    style={
                      Object {
                        "marginRight": undefined,
                      }
                    }
                  >
                    最近创建的话题
                  </div>
                  <div
                    aria-disabled="false"
                    aria-selected="false"
                    className=" ant-tabs-tab"
                    onClick={[Function]}
                    role="tab"
                    style={
                      Object {
                        "marginRight": undefined,
                      }
                    }
                  >
                    最近参与的话题
                  </div>
                  <div
                    aria-disabled="false"
                    aria-selected="false"
                    className=" ant-tabs-tab"
                    onClick={[Function]}
                    role="tab"
                    style={
                      Object {
                        "marginRight": undefined,
                      }
                    }
                  >
                    收藏的话题
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="ant-tabs-content ant-tabs-content-animated"
          style={
            Object {
              "marginLeft": "0%",
            }
          }
        >
          <div
            aria-hidden="false"
            className="ant-tabs-tabpane ant-tabs-tabpane-active"
            role="tabpanel"
          />
          <div
            aria-hidden="true"
            className="ant-tabs-tabpane ant-tabs-tabpane-inactive"
            role="tabpanel"
          />
          <div
            aria-hidden="true"
            className="ant-tabs-tabpane ant-tabs-tabpane-inactive"
            role="tabpanel"
          />
        </div>
      </div>
    </div>
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <div
        className="ant-list ant-list-split"
      >
        <div
          className="ant-spin-nested-loading"
          style={null}
        >
          <div
            className="ant-spin-container"
          >
            <div
              className="ant-list-item"
            >
              <div
                className="ant-list-item-meta"
              >
                <div
                  className="ant-list-item-meta-avatar"
                >
                  <span
                    className="ant-avatar ant-avatar-circle ant-avatar-image"
                  >
                    <img
                      onError={[Function]}
                      src="//"
                    />
                  </span>
                </div>
                <div
                  className="ant-list-item-meta-content"
                >
                  <h4
                    className="ant-list-item-meta-title"
                  >
                    <a
                      href="/topic/007"
                      onClick={[Function]}
                    >
                      title
                    </a>
                  </h4>
                  <div
                    className="ant-list-item-meta-description"
                  >
                    <div>
                      来自：
                      <div
                        className="ant-tag ant-tag-purple"
                        data-show={true}
                        style={
                          Object {
                            "backgroundColor": null,
                          }
                        }
                      >
                        admin
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="ant-list-item-content"
              >
                a day ago
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
